src/Functions/array/FunctionArrayMapped.h

```diff
@@ -219,17 +231,65 @@ class FunctionArrayMapped : public IFunction
                                                           recursiveRemoveLowCardinality(array_type->getNestedType()),
                                                           array_with_type_and_name.name));
             }
+            if (Impl::isFolding())
+                arrays.emplace_back(arguments.back());
 
-            /// Put all the necessary columns multiplied by the sizes of arrays into the columns.
-            auto replicated_column_function_ptr = IColumn::mutate(column_function->replicate(column_first_array->getOffsets()));
-            auto * replicated_column_function = typeid_cast<ColumnFunction *>(replicated_column_function_ptr.get());
-            replicated_column_function->appendArguments(arrays);
+            if (Impl::isFolding() && (! column_first_array->getData().empty()))
+            {
+                size_t arr_cursor = 0;
+                MutableColumnPtr result = arguments.back().column->convertToFullColumnIfConst()->cloneEmpty();
+                for (size_t irow = 0; irow < column_first_array->size(); ++irow) // for each row of result
+                {
+                    // Make accumulator column for this row
+                    ColumnWithTypeAndName accumulator_column = arguments.back();
+                    ColumnPtr acc(accumulator_column.column->cut(irow, 1));
+                    auto accumulator = ColumnWithTypeAndName(acc,
+                                                             accumulator_column.type,
+                                                             accumulator_column.name);
+
+                    ColumnPtr res(acc);
+                    size_t const arr_next = column_first_array->getOffsets()[irow]; // when we do folding
+                    for (size_t iter = 0; arr_cursor < arr_next; ++iter, ++arr_cursor)
+                    {
+                        // Make slice of input arrays and accumulator for lambda
+                        ColumnsWithTypeAndName iter_arrays;
+                        iter_arrays.reserve(arrays.size() + 1);
+                        for (size_t icolumn = 0; icolumn < arrays.size() - 1; ++icolumn)
+                        {
+                            auto const & arr = arrays[icolumn];
+                            iter_arrays.emplace_back(ColumnWithTypeAndName(arr.column->cut(arr_cursor, 1),
```

Comment:
Why do we need this cut? It probably involves allocating memory and copying the data, which is bad for performance. Is it possible to access the input column directly by offsets instead?

Reply:
Can we make sliced copy of ColumnArray by size and offset? I don't find a way without copy, so I decided to use `cut()`. Can you please point me to function to make such slice?

Reply:
It's not possible without a copy so the `cut` method is what you need.